pkg/server/options.go

```diff
@@ -0,0 +1,53 @@
+package server
+
+import (
+	"context"
+	"flag"
+
+	"github.com/spf13/pflag"
+	"k8s.io/klog/v2"
+)
+
+// Options stores option for the command
+type Options struct {
+	// kubeconfig is the path to a KubeConfig file.
+	Kubeconfig string
+	// master is used to override the kubeconfig's URL to the apiserver
+	master           string
+	hostnameOverride string
+	networkPlugins   []string
+	podRulesPath     string
+	// ctx is the execution context, when done, server should stop
+	ctx context.Context
```

Comment:
I think you should not use the same context everywhere. With current implementation, Options.Run method will complete immediately after context is canceled, internal goroutines in the sever may shutoff later.

For me it looks a bit strange that Options struct spawn the server (I know that this approach is used in the iptables based implementation)

I would prefer to change code a bit and do the following in the main.go:
opts := server.NewOptions()
server := server.NewServer(opts)

```
Run: func(cmd *cobra.Command, args []string) {
			if err := server.Run(ctx); err != nil {
				klog.Exit(err)
			}
		},
```

In the server struct you can create internal context, which you will use for all goroutines.
```
func (s *Server) Run(ctx context.Context) {
   s.internalCtx, cFunc := ctx.WithCancel(context.Background())

   <-ctx.Done()
   cFunc()
   // here it will be good to wait for all goroutines to shutdown, 
   // but it looks like it is impossible in our case, so just return
}
```

Reply:
ack will adjust